### PR TITLE
fix: 2d betabinom working

### PR DESCRIPTION
**Pull request by kellino**
Get 2d betabinomial sampling working for image files

### CODE REVIEW

File: rex_xai/distributions.py
Comment (liz-is) [inline rex_xai/distributions.py:68]: `args[0]` is the space/area of the box that's being split, right? It might be good to add a warning here - it's unlikely to be relevant to users, but for people reading/modifying the code, clarifying that this condition means the box is too small to split would make the code clearer.

Alternatively/additionally, it would be helpful to add a note in the docstring on the values expected in args - i.e. it's expected to be space, height, width, then dist_args if present. This would also help make the code easier to follow, without needing to refer to where random_coords() is called.

File: rex_xai/distributions.py
Comment (liz-is) [inline rex_xai/distributions.py:69]: I realise this is following how the existing code works, so feel free to disregard this comment! But I generally prefer not to return a "magic number" like this, as it can be unclear how this should be interpreted. In this case maybe returning `None` would be suitable, looking at how the return value is used in `__2d_parts`. Although a `-1` return value looks like a usable/valid value for `__1d_parts`, so perhaps it is better to return that from this function rather than adding a check for `None` in `__1d_parts`.

**Review by liz-is — APPROVED** [review]
Looks good to me, and I confirmed that it works and gives sensible results on a couple of input images.

I left a couple of comments, but nothing that would block merging. 

I did also notice, while testing, that if I set `dist_args` in the toml file to something that contains one int and one float, I get an error (I would have expected the int to be coerced to a float and not cause issues). But the error seems to have *nothing* to do with the `dist_args`, so I'm not 100% sure what the underlying cause is 😬 Can you reproduce this, or is this just me?

```
(.venv) k2474365@KCLH3912MDY44 ReX % rex rex-repo/imgs/dog.jpg --model models/resnet50-v1-7.onnx -c rex-betabinom.toml
Traceback (most recent call last):
  File "/Users/k2474365/Documents/Projects/Chockler/ReX/.venv/bin/rex", line 6, in <module>
    sys.exit(main())
             ^^^^^^
  File "/Users/k2474365/Documents/Projects/Chockler/ReX/rex-repo/rex_xai/rex.py", line 19, in main
    _ = explanation(args, device)
        ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/k2474365/Documents/Projects/Chockler/ReX/rex-repo/rex_xai/explanation.py", line 230, in explanation
    ps = get_prediction_function(args.model, args.gpu)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/k2474365/Documents/Projects/Chockler/ReX/rex-repo/rex_xai/onnx.py", line 128, in get_prediction_function
    _, ext = os.path.splitext(os.path.basename(model_path))
                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen posixpath>", line 171, in basename
TypeError: expected str, bytes or os.PathLike object, not NoneType
```

**Comment by kellino** [comment]
> Looks good to me, and I confirmed that it works and gives sensible results on a couple of input images.
> 
> I left a couple of comments, but nothing that would block merging.
> 
> I did also notice, while testing, that if I set `dist_args` in the toml file to something that contains one int and one float, I get an error (I would have expected the int to be coerced to a float and not cause issues). But the error seems to have _nothing_ to do with the `dist_args`, so I'm not 100% sure what the underlying cause is 😬 Can you reproduce this, or is this just me?
> 
> ```
> (.venv) k2474365@KCLH3912MDY44 ReX % rex rex-repo/imgs/dog.jpg --model models/resnet50-v1-7.onnx -c rex-betabinom.toml
> Traceback (most recent call last):
>   File "/Users/k2474365/Documents/Projects/Chockler/ReX/.venv/bin/rex", line 6, in <module>
>     sys.exit(main())
>              ^^^^^^
>   File "/Users/k2474365/Documents/Projects/Chockler/ReX/rex-repo/rex_xai/rex.py", line 19, in main
>     _ = explanation(args, device)
>         ^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/Users/k2474365/Documents/Projects/Chockler/ReX/rex-repo/rex_xai/explanation.py", line 230, in explanation
>     ps = get_prediction_function(args.model, args.gpu)
>          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/Users/k2474365/Documents/Projects/Chockler/ReX/rex-repo/rex_xai/onnx.py", line 128, in get_prediction_function
>     _, ext = os.path.splitext(os.path.basename(model_path))
>                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "<frozen posixpath>", line 171, in basename
> TypeError: expected str, bytes or os.PathLike object, not NoneType
> ```

Yes, I can reproduce this, and it's absolutely bizarre! I'll have a look at this

File: rex_xai/distributions.py
Comment (kellino) [reply rex_xai/distributions.py:69]: I've no horror of magic numbers (very C!) but we should at least be consistent. The only reason I prefer -1 is that Python returns None from functions which do not have a formal return, whereas it'll never return -1 accidentally. It's a certain signal, rather than a programming error. We could have an enum for ReXError which would be unambiguous

**Comment by liz-is** [comment]
Ah, got it! I was passing in an onnx model file as a command line argument when I tested, but I guess that fails because if `args` is `None`, `shared_args()` doesn't work correctly.

Btw, you merged before I could comment, but you left in a couple of print statements!

File: rex_xai/distributions.py
Comment (liz-is) [reply rex_xai/distributions.py:69]: Fair enough, that reasoning makes sense to me! I like the idea of defining a ReXError - I tend to default to raising `ValueError` by default but that's not ideal.

And just to confirm, the -1 value is intended to be used in `__1d_parts`? There's no check for it there

**Comment by kellino** [comment]
@liz-is oops to the print statements! Sorry. Shall I fix, or will you?

File: rex_xai/distributions.py
Comment (kellino) [reply rex_xai/distributions.py:69]: I can't remember about the -1 in 1d_parts. I'll have to think about it. We should definitely have a ReXError value. Happy to do this if you want, otherwise I'll leave it to you

**Pull request merged**
